README.md
# AkumaKodo

A blazing fast discord bot framework built on top of [discordeno](https://github.com/discordeno/discordeno) and [typescript](https://www.typescriptlang.org/)!

## Installation

[Click me](https://akumakodo.github.io/AkumaKodo/index.html#installation)

## Example bot

```ts
import { AkumaKodoBotCore } from "https://deno.land/x/akumakodo@0.1.6/mod.ts";
import { config as dotEnvConfig } from "https://deno.land/x/dotenv@v3.1.0/mod.ts";

const env = dotEnvConfig({ export: true });
const TOKEN = env.DISCORD_BOT_TOKEN || "";

// Bot configuration
const Bot = new AkumaKodoBotCore(
    {
        botId: BigInt("your-bot-id"),
        events: {},
        intents: ["Guilds", "GuildMessages", "GuildMembers"],
        token: TOKEN,
    },
    {
        optional: {
            // False by default but we recommend using the built
            // in logging system until your app hits production.
            bot_debug_mode: true,
            // Enables the built in events in the framework
            bot_internal_events: {
                // Allows the bot to reply to slash commands
                interactionCreate: true,
            },
        },
        required: {
            // needed if you wish to use test slash commands.
            bot_development_server_id: BigInt("your-guild-id"),
        },
    },
);

// Our own custom event handler
Bot.client.events.ready = (_, payload) => {
    Bot.container.logger.log("info", "ready", "Online and ready to work!");
};

// Creates a command
Bot.container.utils.createCommand(Bot, {
    trigger: "ping",
    description: "ping pong",
    scope: "Development",
    run: async (interaction) => {
        await Bot.container.utils.createCommandReply(
            Bot,
            interaction,
            {
                embeds: [
                    Bot.container.utils.embed().setColor("random")
                        .setDescription("pong!"),
                ],
            },
            false,
        );
    },
});

// Creates ws connection and starts listening
await Bot.createBot();
```

## Read the [documentation](https://akumakodo.github.io/AkumaKodo/) here!

### Change Log

The current can be found [here](https://akumakodo.github.io/AkumaKodo/misc/changelog.html)

Join my discord [@link](https://discord.com/invite/N79DZsm3m2) for support and feedback!

CHANGELOG.md
# AkumaKodo

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## 0.1.6 - 2022-03-28

Stable release for production.

### Added

- Updated documentation for v0.1.6
- Support team only command checks.

### Changed

- Added optional slash command updater

### Fixed

- More export options to mod.ts file
- Cool down handler bug with not replying to commands.

## 0.1.4 - 2022-03-27

Finishing up basic framework features, and making it stable for production.

### Added

- Command guild only protection.
- Command DM only protection.
- Components for commands: Button, TextInput, SelectMenu.
- More utility helper functions
- Application component handler

### Changed

- Added error handling to createBot function.
- Provider object no longer needed in config by default.
- Providers no longer automatically register using the bot configuration.

### Deprecated

- Removed provider configurations from the bot object. From now on users will have to import and manage their own providers.

### Removed

- Button creation utility.
- Task initialization on bot startup by default.
- Mongodb provider until i fix it.

### Fixed

- Owner Only command checks not caching owners from configuration or api.
- Provider checking for provider object before launching bot.
- Owner Only function working!

### Security

- Upgraded required deno version to deno v1.20.3

## 0.1.3 - 2022-03-25

Major revamp fixes for library and modules with more permission handling.

### Added

- NSFW Command checks
- Owner Only command checks
- development server only checks
- Documentation page for begging users.
- Rate limit handling for commands.

### Changed

- Configuration object added 'required' to props. This will make it more clear to the end user which options are required or optional.
- Bot configuration options have not been given an option to enable and disable built in events
- Version control function name from validate to validateDenoVersion.

### Fixed

- Bot will fetch application owners on start up and cache them.
- Permission handling for event handler. Events are enabled by default and the user can disable them if needed.
- Converted cooldown type names to rate limit for convenience, and understanding.

### Security

- Added more permission checks and caching of the bot owner so we dont hit discord api ratelimit fetch request.

## 0.1.2 - 2022-03-24

Quality of life improvement and bug fixes

### Added

- Error handling for more modules.
- CommandScopeType for command scope options.

### Deprecated

- Removed command scope type Guild in favor of type Developments

### Removed

- The internal events method is now becoming private. To avoid file import bugs, we will load it internally if the user wants to use it. If not they can disable it in the config settings.

### Fixed

- Bug in logger emitting errors for unstable logs.
- The internal event handler was made to accept asynchronous functions.
- framework install url in documentation was incorrect.
- Bug with fileloader import paths

### Security

- If fs loader throws an error, we will end the bot process and not continue loading the process.

## 0.1.1 - 2022-03-22

Quality of life improvements and documentation changes

### Added

- Yoki section to the docs.

### Changed

- Exporting only required classes from mod.ts and not everything.
- base urls for documentation not start with index.md and not a topic name.

### Fixed

- A spelling mistake in the documentation.

## 0.1.0 - 2022-03-19

Initial release of AkumaKodo

### Added

- Interaction Command handling
- Lots of util functions

### Changed

- Added internal slash command handling using events. You can now enable or disable this.
- Logger Errors to warn on un-stable modes.

### Removed

- For now, until slash commands are fully implemented, I will not focus on message-based commands. You will have to implement them yourself.


scripts/clg.js
import {
    Changelog,
    Release,
} from "https://deno.land/x/changelog@v2.0.0/mod.ts";

/** Util to convert js into markdown for our change log files
 *
 * ? Options ~ Added, Changed, Deprecated, Removed, Fixed, and Security
 */
const changelog = new Changelog("AkumaKodo")
    .addRelease(
        new Release("0.1.0", "2022-3-19", "Initial release of AkumaKodo")
            .added("Interaction Command handling")
            .added("Lots of util functions")
            .changed(
                "Added internal slash command handling using events. You can now enable or disable this.",
            )
            .changed("Logger Errors to warn on un-stable modes.")
            .removed(
                "For now, until slash commands are fully implemented, I will not focus on message-based commands. You will have to implement them yourself.",
            ),
    )
    .addRelease(
        new Release(
            "0.1.1",
            "2022-3-22",
            "Quality of life improvements and documentation changes",
        )
            .changed(
                "Exporting only required classes from mod.ts and not everything.",
            )
            .changed(
                "base urls for documentation not start with index.md and not a topic name.",
            )
            .fixed("A spelling mistake in the documentation.")
            .added("Yoki section to the docs."),
    )
    .addRelease(
        new Release(
            "0.1.2",
            "2022-3-24",
            "Quality of life improvement and bug fixes",
        )
            .removed(
                "The internal events method is now becoming private. To avoid file import bugs, we will load it internally if the user wants to use it. If not they can disable it in the config settings.",
            )
            .added("Error handling for more modules.")
            .added("CommandScopeType for command scope options.")
            .fixed("Bug in logger emitting errors for unstable logs.")
            .fixed(
                "The internal event handler was made to accept asynchronous functions.",
            )
            .fixed("framework install url in documentation was incorrect.")
            .fixed("Bug with fileloader import paths")
            .deprecated(
                "Removed command scope type Guild in favor of type Developments",
            )
            .security(
                "If fs loader throws an error, we will end the bot process and not continue loading the process.",
            ),
    )
    .addRelease(
        new Release(
            "0.1.3",
            "2022-3-25",
            "Major revamp fixes for library and modules with more permission handling.",
        )
            .added("NSFW Command checks")
            .fixed(
                "Bot will fetch application owners on start up and cache them.",
            )
            .added("Owner Only command checks")
            .added("development server only checks")
            .added("Documentation page for begging users.")
            .added("Rate limit handling for commands.")
            .changed(
                "Configuration object added 'required' to props. This will make it more clear to the end user which options are required or optional.",
            )
            .changed(
                "Bot configuration options have not been given an option to enable and disable built in events",
            )
            .changed(
                "Version control function name from validate to validateDenoVersion.",
            )
            .fixed(
                "Permission handling for event handler. Events are enabled by default and the user can disable them if needed.",
            )
            .fixed(
                "Converted cooldown type names to rate limit for convenience, and understanding.",
            )
            .security(
                "Added more permission checks and caching of  the bot owner so we dont hit discord api ratelimit fetch request.",
            ),
    )
    .addRelease(
        new Release(
            "0.1.4",
            "2022-3-27",
            "Finishing up basic framework features, and making it stable for production.",
        ).added("Command guild only protection.")
            .added("Command DM only protection.")
            .added("Components for commands: Button, TextInput, SelectMenu.")
            .added("More utility helper functions")
            .added("Application component handler")
            .fixed(
                "Owner Only command checks not caching owners from configuration or api.",
            )
            .changed("Added error handling to createBot function.")
            .changed("Provider object no longer needed in config by default.")
            .fixed(
                "Provider checking for provider object before launching bot.",
            )
            .security("Upgraded required deno version to deno v1.20.3")
            .removed("Button creation utility.")
            .removed("Task initialization on bot startup by default.")
            .fixed("Owner Only function working!")
            .changed(
                "Providers no longer automatically register using the bot configuration.",
            )
            .deprecated(
                "Removed provider configurations from the bot object. From now on users will have to import and manage their own providers.",
            )
            .removed("Mongodb provider until i fix it."),
    ).addRelease(
        new Release(
            "0.1.6",
            "2022-3-28",
            "Stable release for production.",
        ).fixed("More export options to mod.ts file").added(
            "Updated documentation for v0.1.6",
        ).changed("Added optional slash command updater").added(
            "Support team only command checks.",
        ).fixed("Cool down handler bug with not replying to commands."),
    );

console.clear();
console.log("\n- - - COPY THE GENERATED LOG BELOW - - -\n");
console.log(changelog.toString());
console.log("\n- - - COPY THE GENERATED LOG ABOVE - - -\n");
